src/dfx/src/commands/pull.rs

```diff
@@ -114,3 +130,87 @@ async fn fetch_deps_to_pull(
+async fn download_canister_wasm(
+    agent_env: &AgentEnvironment<'_>,
+    logger: &Logger,
+    canister_id: Principal,
+) -> DfxResult {
+    info!(logger, "Downloading wasm of canister {canister_id}...");
+
+    // wasm will be downloaded to $HOME/.cache/dfinity/wasms/{canister_id}/canister.wasm
+    let wasm_dir = get_cache_root()?
+        .join("wasms")
+        .join(canister_id.to_string());
+    let wasm_path = wasm_dir.join("canister.wasm");
+
+    std::fs::create_dir_all(&wasm_dir)
+        .with_context(|| format!("Failed to create dir at {:?}", &wasm_dir))?;
+    // always download and overwrite existing file.
+    let mut wasm_file = std::fs::File::create(&wasm_path)
+        .with_context(|| format!("Failed to create file at {:?}", &wasm_path))?;
+
+    let agent = agent_env
+        .get_agent()
+        .ok_or_else(|| anyhow!("Cannot get HTTP client from environment."))?;
+    let url = match agent
+        .read_state_canister_metadata(canister_id, DFX_WASM_URL)
+        .await
+    {
+        Ok(data) => {
+            let s = String::from_utf8(data)?;
+            reqwest::Url::parse(&s)?
+        }
+        Err(agent_error) => match agent_error {
+            AgentError::HttpError(ref e) => {
+                let content = String::from_utf8(e.content.clone())?;
+                if content.starts_with("Custom section") {
+                    bail!(
+                        "`{}` metadata not found in canister {canister_id}.",
+                        DFX_WASM_URL
+                    );
+                } else {
+                    bail!(agent_error);
+                }
+            }
+            _ => bail!(agent_error),
+        },
+    };
+    let response = reqwest::get(url.clone()).await?;
+    let status = response.status();
+    if status.is_client_error() || status.is_server_error() {
+        bail!("Failed to download wasm from url: {url}.\n  StatusCode: {status}");
+    }
+    let content = response.bytes().await?;
+
+    let canister_status =
+        get_canister_status(agent_env, canister_id, &CallSender::SelectedId).await?;
+    match canister_status.module_hash {
+        Some(hash_on_chain) => {
+            let mut hasher = Sha256::new();
+            hasher.update(&content);
+            let hash_wasm = hasher.finalize();
+            if hash_wasm.as_slice() != hash_on_chain {
+                warn!(
+                    logger,
+                    "The hash of downloaded wasm doesn't match the canister on chain.
+              wasm:     {}
+              on chain: {}",
+                    hex::encode(hash_wasm.as_slice()),
+                    hex::encode(&hash_on_chain)
+                );
+            }
+        }
+        None => {
+            warn!(
+                logger,
+                "Canister {canister_id} doesn't have module hash. Perhaps it's not installed."
+            );
+        }
+    }
+
+    wasm_file.write_all(&content)?;
```

Comment:
It's safer to download (and first write) to a temporary file, and then atomically rename into place.  Otherwise it's possible to have a file with only part of the contents written.